### ISSUE TITLE
docs.rs build failures

### ISSUE THREAD

**Issue by RReverser** [rust-skia/rust-skia#720]
Hi, it looks like the docs.rs build has been failing for all versions starting from 0.18 all the way to latest (0.55): https://docs.rs/crate/skia-safe/latest.

All of them are failing at DNS level for valid URLs:

> Dns Failed: resolve dns name 'github.com:443': failed to lookup address information: Temporary failure in name resolution
> ...
> Dns Failed: resolve dns name 'codeload.github.com:443': failed to lookup address information: Temporary failure in name resolution)

Not sure if this is a failure of docs.rs builder itself, or in rust-skia, but worth investigation and / or reporting upstream?

**Comment by pragmatrix** [comment]
This is a known problem, docs.rs build scripts execute rustdoc in a docker container without network access. I'm planning to detect the `DOCS_RS` environment variable and copy over a pre-generated bindings.rs file with features `gl,textlayout,webp` in the upcoming 0.56 release to see if the docs get rendered.

**Comment by RReverser** [comment]
> This is a known problem, docs.rs build scripts execute rustdoc in a docker container without network access

Huh TIL.

**Comment by pragmatrix** [comment]
Reopening because 0.56 failed:

```
[INFO] [stderr]    Compiling skia-bindings v0.56.0 (/opt/rustwide/workdir)
[INFO] [stderr] error: failed to run custom build command for `skia-bindings v0.56.0 (/opt/rustwide/workdir)`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/skia-bindings-ad0c2506a80e8256/build-script-build` (exit status: 1)
[INFO] [stderr]   --- stdout
[INFO] [stderr]   cargo:rerun-if-env-changed=DOCS_RS
[INFO] [stderr]   DETECTED DOCS_RS BUILD
[INFO] [stderr]   COPYING bindings_docs.rs to src/bindings.rs
[INFO] [stderr] 
[INFO] [stderr]   --- stderr
[INFO] [stderr]   Error: Os { code: 30, kind: ReadOnlyFilesystem, message: "Read-only file system" }
```
This is also related to #10.

**Comment by marc2332** [comment]
Is there any plan to fix this? I can't seem to make the docs build in [freya](https://github.com/marc2332/freya) because of this issue

https://docs.rs/crate/freya/0.1.2/builds/887492

thanks :ok_hand:

**Comment by pragmatrix** [comment]
> Have you considered asking the docs.rs team to make an exception on rust-skia to allow it to run without the offline mode, so it can download the prebuilt binaries?

No, from https://docs.rs/about/builds:

> Network access will not be enabled for any crate.

Since #10 we are referring to `bindings.rs` in `OUT_DIR`, so the strategy originally attempted in #724 may now work. I'll try this with the next release.

**Comment by marc2332** [comment]
Awesome, hopefully it will work :crossed_fingers: 

Any ETA for the next release? No hurries, I just want to be ready :smile: 

Thank you again Armin! :pray:

**Comment by marc2332** [comment]
Hi! it's me again, looks like I found a "working" approach

I have [mocked](https://github.com/marc2332/freya/pull/288/files#diff-90790b7750b97187336411eb59658c0432ce674eae95a7e99d1afd32549a53ef) all the APIs I use from skia-safe, these mocked APIs will only be enabled when building on docs.rs and skia-safe will be disabled. This way it builds on docs.rs!  https://docs.rs/freya/

Thoughts? I know it's ugly but it seems like the best workaround I can do right now

Thanks!

**Comment by pragmatrix** [comment]
I've tried to patch the bindings in with the 0.66.0 release, but failed. The current showstopper here is that with `--allow-dirty`, `cargo publish` won't generate a `.cargo_vcs_info` file and I have not idea how to trick `cargo publish` into allowing a dirty repository _and_ a `.cargo_vcs_info` file. Even if I include the `.cargo_vcs_info` as file to the project, it won't land in the `.crate` file.

@marc2332 I am glad that you found a workaround.

**Comment by marc2332** [comment]
I appreciate the attemps @pragmatrix 😄, thank you. The workaround seems to be good enough for now :v:

_PS: In case you didn't know, I released Freya last week https://marc0.hashnode.dev/freya , it wouldn't have been possible without you 💪_

**Comment by pragmatrix** [comment]
@marc2332 This is awesome, congratulations 🎉 ! And thank you a lot for the credits.

**Comment by tronical** [comment]
Thanks for the release!

It looks like the lack of `.cargo_vcs_info` means that in this release the binaries are never used, because `crate_repository_hash()` fails and therefore `skia-binding`'s `build.rs` fails to determine the correct download url.

**Comment by pragmatrix** [comment]
@tronical  You are right, seems like that `--allow-dirty` slipped through because of all the experiments made. I'll prepare a new release which takes a few hours. Sorry for that.

**Comment by T0mstone** [comment]
This issue still exists (`skia-bindings 0.81.0`), but now I believe the cause is that the generated `bindings_docs.rs` source file isn't included in the published crate on crates.io.

**Comment by T0mstone** [comment]
This issue still exists (skia-bindings 0.81.0), but now I believe the cause is that the generated bindings_docs.rs source file isn't included in the published crate on crates.io.

**Comment by virtualritz** [comment]
Still hitting this on `skia-bindings 0.97.0`. The pre-generated `bindings_docs.rs` is missing from the published tarball even though `Cargo.toml.orig` lists it in `include = [...]`. Reproduction:

```
[INFO] [stderr]    Compiling skia-bindings v0.97.0
[INFO] [stderr] error: failed to run custom build command for `skia-bindings v0.97.0`
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: .../skia-bindings-.../build-script-build (exit status: 1)
[INFO] [stderr]   --- stdout
[INFO] [stderr]   cargo:rerun-if-env-changed=DOCS_RS
[INFO] [stderr]   DETECTED DOCS_RS BUILD
[INFO] [stderr]   COPYING bindings_docs.rs to OUT_DIR/skia/bindings.rs
[INFO] [stderr]   --- stderr
[INFO] [stderr]   Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

(Full log: https://docs.rs/crate/skia-canvas/0.1.0/builds)

Verified the file is missing both from the published 0.97.0 tarball and from the `master` branch on GitHub (404 at `https://raw.githubusercontent.com/rust-skia/rust-skia/master/skia-bindings/bindings_docs.rs`) -- it's generated by the Makefile and apparently not being run before publish.

`skia-safe`'s own docs.rs build shows the same failure: `doc_status: false` on both 0.97.0 and 0.93.1, so every downstream crate that depends on `skia-safe` is also dark on docs.rs.

Opening a PR shortly with a defensive fallback in `fake_bindings()` so this fails gracefully (and ideally regenerates on the fly) instead of bailing the entire build. Will link from here when up.

**Comment by virtualritz** [comment]
PR up: rust-skia/rust-skia#1286 -- root cause turned out to be the master `make publish` target calling `publish-bindings` instead of `publish-bindings-docs`, so the existing docs-prep recipe gets skipped and the file never lands in the tarball. Two-line Makefile fix plus a clearer error message in `fake_bindings()` so future regressions surface a diagnosable error rather than bare `NotFound`.

**Comment by pragmatrix** [comment]
Thank you a lot @virtualritz for following up here, and since https://github.com/rust-lang/cargo/pull/13960, this might actually work now.

**Comment by pragmatrix** [comment]
Just released 0.97.2 and the docs.rs builds went through:

https://docs.rs/skia-bindings/0.97.2
https://docs.rs/skia-safe/0.97.2

**Comment by marc2332** [comment]
amazing! Nice work @virtualritz !
